File: crates/fc-platform/src/webauthn/ceremony_repository.rs
//! WebAuthn ceremony state — short-lived, single-use challenge persistence.
//!
//! Stored in `oauth_oidc_payloads` (the same JSONB store the OIDC provider
//! uses) with type discriminants:
//! - `WebauthnRegistration:{state_id}` — `PasskeyRegistration` blob + principal_id
//! - `WebauthnAuthentication:{state_id}` — `PasskeyAuthentication` blob + optional principal_id
//!
//! `consume_*` uses `DELETE ... RETURNING` so a successful read also marks
//! the state as used in a single round-trip — race-free and replay-safe.

use chrono::{DateTime, Duration, Utc};
use serde_json::json;
use sqlx::PgPool;
use webauthn_rs::prelude::{PasskeyAuthentication, PasskeyRegistration};

use crate::shared::error::{PlatformError, Result};

const REGISTRATION_TYPE: &str = "WebauthnRegistration";
const AUTHENTICATION_TYPE: &str = "WebauthnAuthentication";
const DEFAULT_TTL_SECS: i64 = 600;

fn make_id(kind: &str, state_id: &str) -> String {
    format!("{}:{}", kind, state_id)
}

pub struct WebauthnCeremonyRepository {
    pool: PgPool,
}

pub struct ConsumedRegistration {
    pub principal_id: String,
    pub state: PasskeyRegistration,
    pub display_name: Option<String>,
}

pub struct ConsumedAuthentication {
    pub principal_id: Option<String>,
    pub state: PasskeyAuthentication,
}

impl WebauthnCeremonyRepository {
    pub fn new(pool: &PgPool) -> Self {
        Self { pool: pool.clone() }
    }

    pub async fn store_registration(
        &self,
        state_id: &str,
        principal_id: &str,
        state: &PasskeyRegistration,
        display_name: Option<&str>,
    ) -> Result<()> {
        let payload = json!({
            "principalId": principal_id,
            "displayName": display_name,
            "state": serde_json::to_value(state)
                .map_err(|e| PlatformError::internal(format!("serialise PasskeyRegistration: {}", e)))?,
        });
        let expires_at = Utc::now() + Duration::seconds(DEFAULT_TTL_SECS);
        sqlx::query(
            "INSERT INTO oauth_oidc_payloads (id, type, payload, expires_at, created_at)
             VALUES ($1, $2, $3, $4, NOW())
             ON CONFLICT (id) DO UPDATE SET payload = EXCLUDED.payload, expires_at = EXCLUDED.expires_at",
        )
        .bind(make_id(REGISTRATION_TYPE, state_id))
        .bind(REGISTRATION_TYPE)
        .bind(payload)
        .bind(expires_at)
        .execute(&self.pool)
        .await?;
        Ok(())
    }

    pub async fn consume_registration(
        &self,
        state_id: &str,
    ) -> Result<Option<ConsumedRegistration>> {
        let row: Option<(serde_json::Value,)> = sqlx::query_as(
            "DELETE FROM oauth_oidc_payloads
             WHERE id = $1 AND (expires_at IS NULL OR expires_at > NOW())
             RETURNING payload",
        )
        .bind(make_id(REGISTRATION_TYPE, state_id))
        .fetch_optional(&self.pool)
        .await?;

        let Some((payload,)) = row else {
            return Ok(None);
        };
        let principal_id = payload
            .get("principalId")
            .and_then(|v| v.as_str())
            .ok_or_else(|| PlatformError::internal("ceremony payload missing principalId"))?
            .to_string();
        let display_name = payload
            .get("displayName")
            .and_then(|v| v.as_str())
            .map(String::from);
        let state: PasskeyRegistration = serde_json::from_value(
            payload
                .get("state")
                .cloned()
                .ok_or_else(|| PlatformError::internal("ceremony payload missing state"))?,
        )
        .map_err(|e| PlatformError::internal(format!("deserialise PasskeyRegistration: {}", e)))?;

        Ok(Some(ConsumedRegistration {
            principal_id,
            state,
            display_name,
        }))
    }

    pub async fn store_authentication(
        &self,
        state_id: &str,
        principal_id: Option<&str>,
        state: &PasskeyAuthentication,
    ) -> Result<()> {
        let payload = json!({
            "principalId": principal_id,
            "state": serde_json::to_value(state)
                .map_err(|e| PlatformError::internal(format!("serialise PasskeyAuthentication: {}", e)))?,
        });
        let expires_at = Utc::now() + Duration::seconds(DEFAULT_TTL_SECS);
        sqlx::query(
            "INSERT INTO oauth_oidc_payloads (id, type, payload, expires_at, created_at)
             VALUES ($1, $2, $3, $4, NOW())
             ON CONFLICT (id) DO UPDATE SET payload = EXCLUDED.payload, expires_at = EXCLUDED.expires_at",
        )
        .bind(make_id(AUTHENTICATION_TYPE, state_id))
        .bind(AUTHENTICATION_TYPE)
        .bind(payload)
        .bind(expires_at)
        .execute(&self.pool)
        .await?;
        Ok(())
    }

    pub async fn consume_authentication(
        &self,
        state_id: &str,
    ) -> Result<Option<ConsumedAuthentication>> {
        let row: Option<(serde_json::Value,)> = sqlx::query_as(
            "DELETE FROM oauth_oidc_payloads
             WHERE id = $1 AND (expires_at IS NULL OR expires_at > NOW())
             RETURNING payload",
        )
        .bind(make_id(AUTHENTICATION_TYPE, state_id))
        .fetch_optional(&self.pool)
        .await?;

        let Some((payload,)) = row else {
            return Ok(None);
        };
        let principal_id = payload
            .get("principalId")
            .and_then(|v| v.as_str())
            .map(String::from);
        let state: PasskeyAuthentication = serde_json::from_value(
            payload
                .get("state")
                .cloned()
                .ok_or_else(|| PlatformError::internal("ceremony payload missing state"))?,
        )
        .map_err(|e| {
            PlatformError::internal(format!("deserialise PasskeyAuthentication: {}", e))
        })?;

        Ok(Some(ConsumedAuthentication {
            principal_id,
            state,
        }))
    }

    pub async fn purge_expired(&self) -> Result<u64> {
        let res = sqlx::query(
            "DELETE FROM oauth_oidc_payloads
             WHERE type IN ($1, $2) AND expires_at IS NOT NULL AND expires_at <= NOW()",
        )
        .bind(REGISTRATION_TYPE)
        .bind(AUTHENTICATION_TYPE)
        .execute(&self.pool)
        .await?;
        Ok(res.rows_affected())
    }

    pub fn registration_ttl_seconds(&self) -> DateTime<Utc> {
        Utc::now() + Duration::seconds(DEFAULT_TTL_SECS)
    }
}

File: internal/platform/webauthn/ceremony_repository.go
package webauthn

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/go-webauthn/webauthn/webauthn"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/flowcatalyst/flowcatalyst-go/internal/sqlc/dbq"
)

// CeremonyRepository persists short-lived WebAuthn ceremony state
// (the challenge + session data) between the begin and finish HTTP calls.
//
// Mirrors fc-platform/src/webauthn/ceremony_repository.rs which uses the
// `oauth_oidc_payloads` table for cross-protocol short-lived state.
// Same table here. `Consume*` uses DELETE ... RETURNING so a successful
// read also marks the state as used in a single round-trip — race-free
// and replay-safe.
type CeremonyRepository struct{ q *dbq.Queries }

// NewCeremonyRepository wires a repo.
func NewCeremonyRepository(pool *pgxpool.Pool) *CeremonyRepository {
	return &CeremonyRepository{q: dbq.New(pool)}
}

const (
	registrationType   = "WebauthnRegistration"
	authenticationType = "WebauthnAuthentication"
	defaultTTL         = 10 * time.Minute
)

func makeID(kind, stateID string) string { return kind + ":" + stateID }

// StoreRegistration persists a registration challenge keyed by stateID.
func (r *CeremonyRepository) StoreRegistration(ctx context.Context, stateID, principalID string, session *webauthn.SessionData, displayName *string) error {
	payload, err := json.Marshal(map[string]any{
		"principalId": principalID,
		"session":     session,
		"displayName": displayName,
	})
	if err != nil {
		return err
	}
	expiresAt := time.Now().Add(defaultTTL).UTC()
	return r.q.WebauthnCeremonyUpsert(ctx, dbq.WebauthnCeremonyUpsertParams{
		ID:        makeID(registrationType, stateID),
		Type:      registrationType,
		Payload:   payload,
		ExpiresAt: &expiresAt,
	})
}

// ConsumedRegistration is the registration state recovered by ConsumeRegistration.
type ConsumedRegistration struct {
	PrincipalID string
	Session     webauthn.SessionData
	DisplayName *string
}

// ConsumeRegistration deletes the state with the supplied stateID and
// returns its contents. Returns nil if the state is absent or expired.
func (r *CeremonyRepository) ConsumeRegistration(ctx context.Context, stateID string) (*ConsumedRegistration, error) {
	payload, err := r.q.WebauthnCeremonyConsume(ctx, makeID(registrationType, stateID))
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("consume registration: %w", err)
	}
	var data struct {
		PrincipalID string                `json:"principalId"`
		Session     *webauthn.SessionData `json:"session"`
		DisplayName *string               `json:"displayName"`
	}
	if err := json.Unmarshal(payload, &data); err != nil {
		return nil, err
	}
	if data.Session == nil {
		return nil, errors.New("ceremony state missing session")
	}
	return &ConsumedRegistration{
		PrincipalID: data.PrincipalID,
		Session:     *data.Session,
		DisplayName: data.DisplayName,
	}, nil
}

// StoreAuthentication persists an authentication challenge.
func (r *CeremonyRepository) StoreAuthentication(ctx context.Context, stateID string, principalID *string, session *webauthn.SessionData) error {
	payload, err := json.Marshal(map[string]any{
		"principalId": principalID,
		"session":     session,
	})
	if err != nil {
		return err
	}
	expiresAt := time.Now().Add(defaultTTL).UTC()
	return r.q.WebauthnCeremonyUpsert(ctx, dbq.WebauthnCeremonyUpsertParams{
		ID:        makeID(authenticationType, stateID),
		Type:      authenticationType,
		Payload:   payload,
		ExpiresAt: &expiresAt,
	})
}

// ConsumedAuthentication is the authentication state recovered.
type ConsumedAuthentication struct {
	PrincipalID *string
	Session     webauthn.SessionData
}

// ConsumeAuthentication deletes + returns the authentication state.
func (r *CeremonyRepository) ConsumeAuthentication(ctx context.Context, stateID string) (*ConsumedAuthentication, error) {
	payload, err := r.q.WebauthnCeremonyConsume(ctx, makeID(authenticationType, stateID))
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("consume authentication: %w", err)
	}
	var data struct {
		PrincipalID *string               `json:"principalId"`
		Session     *webauthn.SessionData `json:"session"`
	}
	if err := json.Unmarshal(payload, &data); err != nil {
		return nil, err
	}
	if data.Session == nil {
		return nil, errors.New("ceremony state missing session")
	}
	return &ConsumedAuthentication{PrincipalID: data.PrincipalID, Session: *data.Session}, nil
}

// PurgeExpired removes expired ceremony rows. Mirrors Rust's purge_expired.
func (r *CeremonyRepository) PurgeExpired(ctx context.Context) (int64, error) {
	return r.q.WebauthnCeremonyPurgeExpired(ctx, dbq.WebauthnCeremonyPurgeExpiredParams{
		Type:   registrationType,
		Type_2: authenticationType,
	})
}
